rust/kernel/print.rs

```diff
@@ -0,0 +1,775 @@
+    bindings::printk(
+        format_string.as_ptr() as _,
+        string.len() as c_int,
+        string.as_ptr(),
+    );
+}
+
+// From `kernel/print/printk.c`.
+const LOG_LINE_MAX: usize = 1024 - 32;
+
+#[doc(hidden)]
+pub struct LogLineWriter {
+    data: [u8; LOG_LINE_MAX],
+    pos: usize,
+}
+
+impl LogLineWriter {
+    /// Creates a new [`LogLineWriter`].
+    pub fn new() -> LogLineWriter {
+        LogLineWriter {
+            data: [0u8; LOG_LINE_MAX],
+            pos: 0,
+        }
+    }
+
+    /// Returns the internal buffer as a byte slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data[..self.pos]
+    }
+}
+
+impl Default for LogLineWriter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl fmt::Write for LogLineWriter {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        let copy_len = cmp::min(LOG_LINE_MAX - self.pos, s.as_bytes().len());
+        self.data[self.pos..self.pos + copy_len].copy_from_slice(&s.as_bytes()[..copy_len]);
+        self.pos += copy_len;
+        Ok(())
+    }
+}
+
+/// Performs formatting and forwards the string to [`call_printk`].
+///
+/// This is an *unsafe* macro.
+///
+/// Hidden because it is called from a public macro -- it should never be
+/// called by users.
+///
+/// # Safety
+///
+/// See [`call_printk`]'s and [`call_printk_cont`]'s safety contract.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! print_macro (
+    // Without extra arguments: no need to format anything, so simplify.
+    ($format_string:path, $call_fn:ident, $suffix:literal, $fmt:expr) => (
+        // SAFETY: This hidden macro should only be called by the documented
+        // printing macros which ensure the format string is a fixed one.
+        unsafe {
+            $crate::print::$call_fn(
+                &$format_string,
+                __MODULE_NAME,
+                concat!($fmt, $suffix).as_bytes()
+            )
+        };
+    );
+
+    // With extra arguments: we need to perform formatting.
+    ($format_string:path, $call_fn:ident, $suffix:literal, $fmt:expr, $($arg:tt)*) => (
+        // The immediately-invoked closure is needed here to avoid stack
+        // overflows:
+        //
+        //   - In non-optimized builds, without it, the compiler reserves one
+        //     [`LogLineWriter`] per macro invocation, which is a huge type.
+        //
+        //   - We could use a function for this, since the result of
+        //     [`format_args!`] does not need heap allocations and can be
+        //     nicely passed as an argument. However, that also seems to use
+        //     more stack than the immediately-invoked closures.
+        //
+        // The code size increases in non-optimized builds, but at `O2`
+        // optimization levels the generated code is basically the same for
+        // all alternatives.
+        (|| {
+            use ::core::fmt;
+            let mut writer = $crate::print::LogLineWriter::new();
```

Comment:
Using up 1k of the limited stack seems excessive. Can we do better?

Reply:
Yeah, I think eventually we want to have a smarter way. An easy one perhaps is to simply call `printk(` with continuations of a smaller buffer. For the moment, we can reduce the buffer size if needed, e.g. to `300` which should be plenty for the vast majority of things.

In any case, note that this is the current state -- this PR is more about adding the overall API so that we can using it in drivers etc. (actually, this PR improves the situation on the stack blowing up -- please see the other comments).

Reply:
Agreed that it's an improvement.

Could we reduce the size to `300` for now? 1k is 12.5% of all available stack space...

Reply:
Yeah, I think it is a good idea to reduce it -- will do.

Reply:
Done in upcoming push.